core/primitives/src/validator_signer.rs

```diff
@@ -0,0 +1,230 @@
+use std::path::Path;
+use std::sync::Arc;
+
+use borsh::BorshSerialize;
+
+use near_crypto::{InMemorySigner, KeyType, PublicKey, Signature, Signer};
+
+use crate::block::{Approval, BlockHeader, BlockHeaderInnerLite, BlockHeaderInnerRest};
+use crate::challenge::ChallengeBody;
+use crate::hash::{hash, CryptoHash};
+use crate::network::{AnnounceAccount, PeerId};
+use crate::sharding::{ChunkHash, ShardChunkHeaderInner};
+use crate::types::{AccountId, BlockHeight, EpochId};
+
+/// Validator signer that is used to sign blocks and approvals.
+pub trait ValidatorSigner: Sync + Send {
+    /// Account id of the given validator.
+    fn validator_id(&self) -> &AccountId;
+
+    /// Public key that identifies this validator.
+    fn public_key(&self) -> PublicKey;
+
+    /// Sign json (used for info).
+    fn sign_json(&self, json: String) -> Signature;
```

Comment:
Sadly it doesn't really help with security because "signer" still signs json directly (e.g. HSM will receive JSON).
Which means to clean up this interface we actually need to move serialization inside this method and return serialized object.